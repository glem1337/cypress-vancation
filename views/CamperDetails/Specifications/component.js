import { Button, Col, Row, Tag } from 'antd';
import { FormattedMessage } from 'react-intl';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { MAX_DESCRIPTION_VISIBLE_SYMBOLS } from 'constants/camperDetails';

import isPresent from 'utils/isPresent';

import TooltipIcon from 'views/shared/TooltipIcon';
import FormattedOrRawMessage from 'views/shared/FormattedOrRawMessage';
import SkeletonText from 'views/shared/SkeletonText';
import LOCATORS from 'constants/locators';

const Specifications = ({
  camper,
  allDescriptionVisible,
  allDetailsVisible,
  toggleDescription,
  toggleDetails,
  description,
  detailsConfig,
  isCamperExist,
}) => {
  if (!isCamperExist || !isPresent(camper.specificationDetail)) {
    return (
      <>
        <div className="skeleton__title w-100" />
        <SkeletonText rows={4} />
      </>
    );
  }

  const { whoBuiltCamper, ...details } = camper.specificationDetail;

  return (
    <>
      <div id="details-overview" className="van-details__info">
        <div className="text-headline mr-16">
          <FormattedMessage id="shared.camperInfo" />
        </div>
        <div className="van-details__info-tag-wrap">
          <Tag className="mr-8" color="default">
            <span className="font-14">
              <span className="in-black">
                <FormattedMessage id="shared.builtBy" />
                :
              </span>
              &nbsp;
              <span className="in-blue-1000 font-600">{whoBuiltCamper}</span>
            </span>
          </Tag>
        </div>
        <p className="van-details__info-numb">
          <FormattedMessage id="shared.rental" />
          {' '}
          #
          {camper.publicId}
        </p>
      </div>
      {isPresent(camper.description) && (
        <div className="mb-24">
          <p className="text-subheader font-400 text-color-gray">
            {description}
          </p>
          {camper.description.length > MAX_DESCRIPTION_VISIBLE_SYMBOLS && (
            <Button
              onClick={toggleDescription}
              size="small"
              className="mt-4 text-subheader font-400 text-color-gray h-auto p-0"
              type="simple-text"
            >
              <span>
                <FormattedMessage
                  id={allDescriptionVisible ? 'shared.hide' : 'shared.readMore'}
                />
              </span>
            </Button>
          )}
        </div>
      )}
      <div className="mb-24 mb-md-40">
        <p className="text-subheader mb-16 font-700">
          <FormattedMessage id="shared.details" />
        </p>
        <Row data-targetId="camper-info-list" gutter={24}>
          {detailsConfig.map((item) => {
            let config = {
              icon: `/images/${item.icon}.svg`,
              value: details[item.id] || 0,
            };

            if (isPresent(item.config)) {
              config = {
                ...item.config[details[item.id]],
              };
            }

            return (
              <Col key={item.id} md={8} xl={6}>
                <div className="van-details__details-card">
                  <div className="van-details__details-card-img">
                    <img src={config.icon} alt="" />
                  </div>
                  <div className="mr-auto">
                    <p>
                      <FormattedMessage {...item.title} />
                    </p>
                    <p
                      className={classNames(
                        'text-subheader font-700',
                        item.classes && item.classes,
                      )}
                    >
                      <FormattedOrRawMessage message={config.value} />
                      {item.dimensions && (
                        <>
                          {' '}
                          <FormattedMessage {...item.dimensions} />
                        </>
                      )}
                    </p>
                  </div>
                  {item.tooltip && (
                    <TooltipIcon
                      phrase={<FormattedMessage {...item.tooltip} />}
                    />
                  )}
                </div>
              </Col>
            );
          })}
        </Row>
        <button
          onClick={toggleDetails}
          type="button"
          className="in-blue-1000 main-link font-600"
          data-targetId={LOCATORS.CAMPER_DETAILS.CAMPER_INFO.BUTTON}
        >
          <FormattedMessage
            id={allDetailsVisible ? 'shared.hide' : 'shared.showAll'}
          />
        </button>
      </div>
    </>
  );
};

Specifications.defaultProps = {
  camper: null,
};

Specifications.propTypes = {
  camper: PropTypes.shape(),
  isCamperExist: PropTypes.bool.isRequired,
  allDescriptionVisible: PropTypes.bool.isRequired,
  allDetailsVisible: PropTypes.bool.isRequired,
  toggleDescription: PropTypes.func.isRequired,
  toggleDetails: PropTypes.func.isRequired,
  description: PropTypes.string.isRequired,
  detailsConfig: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default Specifications;
