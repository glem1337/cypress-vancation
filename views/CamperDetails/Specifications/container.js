import React from 'react';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import { connect } from 'react-redux';

import {
  DETAILS_CONFIG,
  MAX_DESCRIPTION_VISIBLE_SYMBOLS,
  DEFAULT_VISIBLE_DETAILS,
} from 'constants/camperDetails';

import isMobileView from 'utils/breakpoints/isMobileView';
import isTabletView from 'utils/breakpoints/isTabletView';

import {
  camperSelector,
  isCamperExistSelector,
} from 'state/concepts/camper/selectors';

import SpecificationsComponent from './component';

class Specifications extends React.Component {
  static defaultProps = {
    camper: null,
  };

  static propTypes = {
    camper: PropTypes.shape(),
  };

  constructor(props) {
    super(props);

    this.state = {
      allDescriptionVisible: false,
      allDetailsVisible: false,
    };
  }

  toggleDescription = () => {
    this.setState((prevState) => ({
      allDescriptionVisible: !prevState.allDescriptionVisible,
    }));
  };

  get description() {
    const { camper } = this.props;
    const { allDescriptionVisible } = this.state;

    const description = R.pathOr('', ['description'], camper);

    if (
      allDescriptionVisible
      || description.length <= MAX_DESCRIPTION_VISIBLE_SYMBOLS
    ) {
      return description;
    }

    return `${description.substring(0, MAX_DESCRIPTION_VISIBLE_SYMBOLS)}...`;
  }

  toggleDetails = () => {
    this.setState((prevState) => ({
      allDetailsVisible: !prevState.allDetailsVisible,
    }));
  };

  get detailsConfig() {
    const { allDetailsVisible } = this.state;

    const isDesktop = !isMobileView() && !isTabletView();
    const itemsCount = isDesktop
      ? DEFAULT_VISIBLE_DETAILS.DESKTOP
      : DEFAULT_VISIBLE_DETAILS.MOBILE;

    return allDetailsVisible
      ? DETAILS_CONFIG
      : DETAILS_CONFIG.slice(0, itemsCount);
  }

  render() {
    return (
      <SpecificationsComponent
        {...this.state}
        {...this.props}
        toggleDescription={this.toggleDescription}
        description={this.description}
        detailsConfig={this.detailsConfig}
        toggleDetails={this.toggleDetails}
      />
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  camper: camperSelector(state, ownProps?.camperId),
  isCamperExist: isCamperExistSelector(state, ownProps?.camperId),
});

export { Specifications as SpecificationsContainer };
export default connect(mapStateToProps)(Specifications);
