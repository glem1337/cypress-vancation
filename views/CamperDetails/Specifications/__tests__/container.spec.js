import { shallow } from 'enzyme';
import * as R from 'ramda';
import configureStore from 'redux-mock-store';

import {
  DEFAULT_VISIBLE_DETAILS,
  DETAILS_CONFIG,
  MAX_DESCRIPTION_VISIBLE_SYMBOLS,
} from 'constants/camperDetails';

import diveTo from 'utils/testHelpers/diveToEnzyme';

import mockedCamper from 'views/__mocks__/camper';
import Specifications, { SpecificationsContainer } from '../container';

jest.mock('utils/breakpoints/isMobileView', () => jest.fn(() => false));
jest.mock('utils/breakpoints/isTabletView', () => jest.fn(() => false));

jest.mock('state/concepts/camper/selectors', () => ({
  camperSelector: jest.fn(() => mockedCamper),
  isCamperExistSelector: jest.fn(() => true),
}));

describe('Specifications container tests', () => {
  const store = configureStore()({});
  store.dispatch = jest.fn();

  const props = {
    store,
  };

  let wrapper = null;
  let container = null;
  let instance = null;

  beforeEach(() => {
    wrapper = shallow(<Specifications {...props} />);
    container = diveTo(wrapper, SpecificationsContainer);
    instance = container.instance();

    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  it('checks `toggleDescription` instance method', () => {
    instance.toggleDescription();

    expect(container.state().allDescriptionVisible).toBe(true);

    instance.toggleDescription();

    expect(container.state().allDescriptionVisible).toBe(false);
  });

  describe('checks `description` instance getter', () => {
    it('should return not truncated description', () => {
      const expected = R.pathOr('', ['description'], mockedCamper);

      expect(instance.description).toEqual(expected);
    });

    it('should return truncated description', () => {
      container.setProps({
        ...props,
        camper: {
          ...mockedCamper,
          description:
            "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essenti",
        },
      });

      const { camper } = container.props();

      const description = R.pathOr('', ['description'], camper);

      const expected = `${description.substring(
        0,
        MAX_DESCRIPTION_VISIBLE_SYMBOLS,
      )}...`;

      expect(instance.description).toBe(expected);
    });
  });

  it('checks `toggleDetails` instance method', () => {
    instance.toggleDetails();

    expect(container.state().allDetailsVisible).toBe(true);

    instance.toggleDetails();

    expect(container.state().allDetailsVisible).toBe(false);
  });

  it('checks `detailsConfig` instance getter', () => {
    const expected = DETAILS_CONFIG.slice(0, DEFAULT_VISIBLE_DETAILS.DESKTOP);

    expect(instance.detailsConfig).toEqual(expected);
  });
});
