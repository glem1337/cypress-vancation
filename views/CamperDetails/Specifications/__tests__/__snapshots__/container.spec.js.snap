// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Specifications container tests matches snapshot 1`] = `
<Specifications
  allDescriptionVisible={false}
  allDetailsVisible={false}
  camper={
    Object {
      "blockedPeriods": Array [],
      "camperAddition": Object {
        "amenityHealthSafetyItems": Array [
          Object {
            "healthSafety": Object {
              "iconUrl": "path/to/icon/url",
              "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
              "name": "name",
            },
          },
        ],
        "camperQuestions": Array [
          Object {
            "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
            "required": true,
            "text": "test",
          },
        ],
        "customRestrictionRoads": Array [
          Object {
            "active": true,
            "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
            "name": "name",
          },
        ],
        "customRestrictionRules": Array [
          Object {
            "active": true,
            "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
            "name": "name",
          },
        ],
        "customTravelAccessories": Array [
          Object {
            "active": true,
            "description": "",
            "id": "dd477886-ec22-4da9-ba4c-08627f35fb1a",
            "maxAmount": "2.0",
            "name": "Bike Rack",
            "price": "1.0",
            "priceUnit": "each",
          },
        ],
        "customTravelRestrictions": Array [
          Object {
            "active": true,
            "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
            "name": "name",
          },
        ],
        "restrictionRoad": Object {
          "dirtryRoad": true,
          "fourWheelRoad": false,
          "offRoad": false,
          "snowAndIceRoad": false,
        },
        "restrictionRule": Object {
          "allowPets": false,
          "festivalApproved": false,
          "smoking": false,
        },
        "travelRestriction": Object {
          "burningMan": false,
          "canada": true,
          "mexico": false,
        },
      },
      "camperCalendar": Object {
        "blockedPeriods": Array [
          Object {
            "endDate": "2021-08-06",
            "id": "56ceb535-322f-42c6-8dcd-a88ad59000f3",
            "startDate": "2021-08-05",
          },
          Object {
            "endDate": "2021-08-25",
            "id": "17a32ec0-a486-4247-910c-6c0c57ba6ec3",
            "startDate": "2021-08-25",
          },
          Object {
            "endDate": "2021-07-28",
            "id": "2c7bbaef-6b00-494f-b005-6a762e90c785",
            "startDate": "2021-07-28",
          },
        ],
        "customDiscountPeriods": Array [
          Object {
            "endDate": "2021-07-28",
            "id": "ffb3f0cc-4513-4fdc-81d0-03587232d9ee",
            "monthlyDiscountPercent": 23,
            "startDate": "2021-07-27",
            "weeklyDiscountPercent": 15,
          },
          Object {
            "endDate": "2021-07-28",
            "id": "a070339d-7058-4769-8e16-814084c6bfe3",
            "monthlyDiscountPercent": 23,
            "startDate": "2021-07-28",
            "weeklyDiscountPercent": 15,
          },
          Object {
            "endDate": "2021-07-28",
            "id": "aba16cba-5873-4273-8172-816b276a2db7",
            "monthlyDiscountPercent": 44,
            "startDate": "2021-07-28",
            "weeklyDiscountPercent": 44,
          },
          Object {
            "endDate": "2021-07-29",
            "id": "586016b9-d351-4ca1-ac8a-373e7e3b8832",
            "monthlyDiscountPercent": 33,
            "startDate": "2021-07-29",
            "weeklyDiscountPercent": 15,
          },
        ],
        "externalCalendars": Array [],
        "id": "e1db81b1-8ad5-4e88-9c99-f5c9020148eb",
        "pricingPeriods": Array [
          Object {
            "endDate": "2021-07-28",
            "id": "8aa92a51-ef00-4dc6-8b84-2820869b9278",
            "price": "33.0",
            "startDate": "2021-07-28",
          },
          Object {
            "endDate": "2021-07-30",
            "id": "248ba38e-e18a-467c-996b-36aa4bd73e85",
            "price": "320.0",
            "startDate": "2021-07-29",
          },
        ],
      },
      "camperPhotos": Array [
        Object {
          "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
          "photoUrl1100": "photoUrl1100",
        },
      ],
      "camperRule": Object {
        "bookingApprovalPolicy": "instant_book",
        "cancellationPolicy": "easy_going",
        "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
      },
      "deliveryInformation": Object {
        "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
        "pickup": true,
      },
      "description": "desc",
      "estimatedEarning": 1200,
      "externalCalendars": Array [],
      "glamper": true,
      "id": "a18f7338-c2b3-4cee-be4d-579c0a726c1a",
      "insurance": "pending",
      "insuranceInfo": Object {
        "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
      },
      "latitude": 40.772,
      "longitude": -73.93030000000002,
      "name": "test",
      "owner": Object {
        "account": Object {
          "createdAt": "2021-07-09T13:39:26.587Z",
          "drivingLicenceVerified": false,
          "email": "kateryna.sessa+owner@rubygarage.org",
          "emailVerified": false,
          "id": "12563773-6e8f-4861-8abd-82a78632c6ac",
          "user": Object {
            "avatarUrl": null,
            "createdAt": "2021-07-09T13:39:26.590Z",
            "firstName": "Kat",
            "id": "77015afb-71e3-4b94-a9cb-69a97dc6b358",
            "lastName": "Kat",
            "updatedAt": "2021-07-09T13:39:26.590Z",
          },
        },
        "avatarUrl": null,
        "averageRating": null,
        "businessTitle": null,
        "campersCount": 25,
        "createdAt": "Member since July 2021",
        "description": "description",
        "email": null,
        "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
        "idVerified": false,
        "type": "owner",
      },
      "place": "Astoria, New York, New York 11102, United States",
      "placeId": "neighborhood.290416",
      "pricingInfo": Object {
        "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
      },
      "pricingPeriods": Array [],
      "publicId": "111",
      "specificationDetail": Object {
        "drivetrain": "4x4",
        "fuelType": "electric",
        "insideHeight": "low_top",
        "length": "67.89",
        "mileage": "Under 50,000",
        "modelNaming": "Clair Weimann",
        "name": "Katelyn Lowe",
        "seats": 1,
        "sleeps": 1,
        "stateRegistred": "Arkansas",
        "transmission": "automatice",
        "whoBuiltCamper": "Vasya",
        "year": 1585,
      },
      "status": "draft",
      "tripFee": Object {
        "cleaning": 1,
        "customFees": Array [
          Object {
            "frequency": "per_day",
            "id": "id",
            "name": "name",
            "price": 5,
          },
        ],
        "dumpingFee": 2,
        "flue": 3,
        "id": "74a9c587-7b78-4aa5-ba93-ba3d310be127",
        "lateDropOff": 4,
        "tripFeeGenerator": Object {
          "available": 8,
          "limit": true,
          "overage": 9,
        },
        "tripFeeMileage": Object {
          "available": 6,
          "limit": true,
          "overage": 7,
        },
      },
      "vehicleTypeIconUrl": "test.svg",
      "vehicleTypeName": "Modern Van",
    }
  }
  description="desc"
  detailsConfig={
    Array [
      Object {
        "icon": "details_camp/Sleeps",
        "id": "sleeps",
        "title": Object {
          "id": "addNewCamper.camperDetails.form.sleeps",
        },
      },
      Object {
        "icon": "details_camp/Seats",
        "id": "seats",
        "title": Object {
          "id": "addNewCamper.camperDetails.form.seats",
        },
        "tooltip": Object {
          "id": "addNewCamper.camperDetails.form.seats",
        },
      },
      Object {
        "config": Object {
          "hight_top": Object {
            "icon": "/images/listing/High-Top.svg",
            "value": Object {
              "id": "addNewCamper.camperDetails.from.highTop",
            },
          },
          "low_top": Object {
            "icon": "/images/listing/Low-Top.svg",
            "value": Object {
              "id": "addNewCamper.camperDetails.from.lowTop",
            },
          },
          "pop_top": Object {
            "icon": "/images/listing/Pop-Top.svg",
            "value": Object {
              "id": "addNewCamper.camperDetails.from.popTop",
            },
          },
          "roof_top": Object {
            "icon": "/images/listing/Roof-Top-Tent.svg",
            "value": Object {
              "id": "addNewCamper.camperDetails.from.roofTopTent",
            },
          },
        },
        "icon": "listing/High-Top",
        "id": "insideHeight",
        "title": Object {
          "id": "addNewCamper.camperDetails.form.insideHeight",
        },
      },
      Object {
        "icon": "details_camp/Length",
        "id": "length",
        "title": Object {
          "id": "addNewCamper.camperDetails.form.length",
        },
      },
      Object {
        "classes": "text-capitalize",
        "icon": "details_camp/Transmission",
        "id": "transmission",
        "title": Object {
          "id": "addNewCamper.camperDetails.form.transmission",
        },
      },
      Object {
        "icon": "edit_listing/rules/Allow-Unlimited-Miles",
        "id": "mileage",
        "title": Object {
          "id": "shared.currentMileage",
        },
      },
      Object {
        "classes": "text-capitalize",
        "icon": "details_camp/Fuel-Type",
        "id": "fuelType",
        "title": Object {
          "id": "addNewCamper.camperDetails.form.fuelType",
        },
      },
      Object {
        "classes": "text-uppercase",
        "icon": "edit_listing/rules/4x4-Only-Roads",
        "id": "drivetrain",
        "title": Object {
          "id": "addNewCamper.camperDetails.form.drivetrain",
        },
      },
    ]
  }
  dispatch={[MockFunction]}
  isCamperExist={true}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  toggleDescription={[Function]}
  toggleDetails={[Function]}
/>
`;
